File: In development/web_api.py
import json
import sqlite3
import subprocess
from pathlib import Path

from fastapi import FastAPI
from fastapi.staticfiles import StaticFiles

from piwardrive.security import validate_service_name

app = FastAPI()


@app.get("/api/gps")
def get_gps():
    """Return the latest GPS entry from the wardrive database."""
    # Replace with actual DB path in production
    db_path = Path("/home/pi/.config/piwardrive/db.sqlite")
    if not db_path.exists():
        return {"lat": None, "lon": None, "time": None}
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute("SELECT lat, lon, time FROM gps_tracks ORDER BY time DESC LIMIT 1")
    row = cursor.fetchone()
    conn.close()
    if row:
        return {"lat": row[0], "lon": row[1], "time": row[2]}
    return {"lat": None, "lon": None, "time": None}


@app.get("/api/aps")
def get_aps():
    """Return saved Wi‑Fi access points as GeoJSON."""
    file = Path("/home/pi/.config/piwardrive/export/aps.geojson")
    if file.exists():
        with file.open() as f:
            return json.load(f)
    return {"type": "FeatureCollection", "features": []}


@app.get("/api/bt")
def get_bt():
    """Return Bluetooth device data as GeoJSON."""
    file = Path("/home/pi/.config/piwardrive/export/bt.geojson")
    if file.exists():
        with file.open() as f:
            return json.load(f)
    return {"type": "FeatureCollection", "features": []}


def _toggle_service(service: str, state: str) -> dict:
    """Start or stop a systemd service."""
    validate_service_name(service)
    if state == "start":
        subprocess.run(["sudo", "systemctl", "start", service], check=False)
    elif state == "stop":
        subprocess.run(["sudo", "systemctl", "stop", service], check=False)
    return {"status": "ok"}


@app.post("/api/service/{service}/toggle")
def toggle_service(service: str, state: str):
    """Toggle an arbitrary service."""
    return _toggle_service(service, state)


@app.post("/api/kismet/toggle")
def toggle_kismet(state: str):
    """Start or stop the kismet service."""
    return _toggle_service("kismet", state)


@app.post("/api/bettercap/toggle")
def toggle_bettercap(state: str):
    """Start or stop the BetterCAP service."""
    return _toggle_service("bettercap", state)


# Serve static dashboard if available
_static = Path(__file__).parent / "web_gui"
if _static.is_dir():
    app.mount("/", StaticFiles(directory=_static, html=True), name="static")

File: In development/web_api.js
const http = require('http');
const fs = require('fs');
const { spawnSync } = require('child_process');
const url = require('url');

function readGps(dbFile) {
  try {
    const res = spawnSync('sqlite3', ['-noheader', '-separator', '|', dbFile,
      'SELECT lat, lon, time FROM gps_tracks ORDER BY time DESC LIMIT 1;']);
    const output = res.stdout.toString().trim();
    if (!output) return null;
    const [lat, lon, time] = output.split('|');
    return { lat: parseFloat(lat), lon: parseFloat(lon), time };
  } catch {
    return null;
  }
}

function readGeo(file) {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
  } catch {
    return { type: 'FeatureCollection', features: [] };
  }
}

function toggleService(service, state) {
  const cmd = state === 'start'
    ? ['sudo', 'systemctl', 'start', service]
    : ['sudo', 'systemctl', 'stop', service];
  if (process.env.PW_RUN_LOG) {
    fs.appendFileSync(process.env.PW_RUN_LOG, JSON.stringify(cmd) + '\n');
  } else {
    spawnSync(cmd[0], cmd.slice(1));
  }
  return { status: 'ok' };
}

function createServer(opts = {}) {
  const gpsDb = opts.gpsDb || process.env.PW_GPS_DB || '/home/pi/.config/piwardrive/db.sqlite';
  const apsFile = opts.apsFile || process.env.PW_APS_FILE || '/home/pi/.config/piwardrive/export/aps.geojson';
  const btFile = opts.btFile || process.env.PW_BT_FILE || '/home/pi/.config/piwardrive/export/bt.geojson';

  return http.createServer((req, res) => {
    const parsed = url.parse(req.url, true);
    res.setHeader('Content-Type', 'application/json');
    if (req.method === 'GET' && parsed.pathname === '/api/gps') {
      res.end(JSON.stringify(readGps(gpsDb) || { lat: null, lon: null, time: null }));
    } else if (req.method === 'GET' && parsed.pathname === '/api/aps') {
      res.end(JSON.stringify(readGeo(apsFile)));
    } else if (req.method === 'GET' && parsed.pathname === '/api/bt') {
      res.end(JSON.stringify(readGeo(btFile)));
    } else if (req.method === 'POST' && parsed.pathname === '/api/kismet/toggle') {
      res.end(JSON.stringify(toggleService('kismet', parsed.query.state)));
    } else {
      res.statusCode = 404;
      res.setHeader('Content-Type', 'text/plain');
      res.end('Not found');
    }
  });
}

if (require.main === module) {
  const port = process.env.PORT !== undefined ? Number(process.env.PORT) : 3000;
  const server = createServer();
  server.listen(port, () => {
    console.log(`Listening on ${server.address().port}`);
  });
}

module.exports = { createServer };
